framework/src/MonteCarloUtility.cu

```diff
@@ -0,0 +1,122 @@
+#include "MonteCarloUtility.h"
+
+#include <ctime>
+
+#include <cuda_runtime.h>
+#include <curand_kernel.h>
+#include <helper_cuda.h>
+
+__device__ __constant__ Player::playernum_t initPlayerTurn;
+__device__ __constant__ Game::GameBoard initalGameBoard;
+__device__ gpu_result gpuResultGlobal;
+__device__ curandState_t curandStatesGlobal[BLOCK_SIZE];
+
+// Kernel to init curand
+__global__ void curandInitKernel(unsigned long long seed)
+{
+    // Init curand states
+    curand_init(seed, threadIdx.x, 0, &curandStatesGlobal[threadIdx.x]);
+}
+
+__global__ void simulationKernel()
+{
+    // Initialize result memory (to be copied out later)
+    // First two are players, third is total play count
+    __shared__ gpu_result gpuResultLocal;
+
+    // Init result to 0
+    if(threadIdx.x == 0)
+        gpuResultLocal = {};
+    __syncthreads();
+
+    // Grab curand state 
+    curandState_t curandStateLocal = curandStatesGlobal[threadIdx.x];
+
+    // Init search states
+    Player::playernum_t currentPlayerTurn = initPlayerTurn;
+    Game::GameBoard currentBoardState = initalGameBoard;
+    Game::boardresult_t currentBoardResult = Game::GAME_ACTIVE;
+
+    // Do simulations until threshold reached
+    while(gpuResultLocal.playCount < PLAYCOUNT_THRESHOLD_GPU)
+    {
+        // Pick random move and execute
+
+        // Get list of possible moves
+        Game::movelist_t moveList;
+        Game::movecount_t moveCount = currentBoardState.getMoves(moveList, currentPlayerTurn);
+        
+        // Prevent floating point exceptions
+        if(moveCount > 0)
+        {   
+            // Select random move
+            Game::move_t selectedMove = moveList[curand(&curandStateLocal) % moveCount];
+
+            // Execute random move
+            Game::moveresult_t moveResult = currentBoardState.executeMove(selectedMove, currentPlayerTurn);
+
+            // Check Move
+            // TODO: Add check to make sure move is not invalid
```

Comment:
Is there a plan to implement this check? Seems like it would be worth doing.

Reply:
Yeah, it probably would be work doing. The question is how it is handled. Do we just print out and continue, do we not print and reset, do we print and exit, etc?

Reply:
one idea would be to not put the check there and just write a unit test to validate that the function returns what it is supposed to.